### PR TITLE
SpecialPower

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/rts/specialpower.cpp [hunk 1]
```diff
@@ -62,6 +62,98 @@ bool SpecialPowerStore::Can_Use_Special_Power(Object *obj, const SpecialPowerTem
...
+const SpecialPowerTemplate *SpecialPowerStore::Get_Special_Power_Template_By_Index(unsigned int index)
+{
+    if (index < m_specialPowerTemplates.size()) {
+        return m_specialPowerTemplates[index];
+    } else {
+        return nullptr;
+    }
+}
+
+int SpecialPowerStore::Get_Num_Special_Powers()
```
Comment (xezon) [inline src/game/common/rts/specialpower.cpp:90]: Not returning `unsigned int`?

File: src/game/common/rts/specialpower.cpp [hunk 2]
```diff
@@ -62,6 +62,98 @@ bool SpecialPowerStore::Can_Use_Special_Power(Object *obj, const SpecialPowerTem
...
+
+        power = new SpecialPowerTemplate();
+        const SpecialPowerTemplate *default_power =
+            g_theSpecialPowerStore->Find_Special_Power_Template("DefaultSpecialPower");
+
+        if (default_power != nullptr) {
+            *power = *default_power;
+        }
+
+        power->Set_Name_ID(str, ++g_theSpecialPowerStore->m_nextSpecialPowerID);
```
Comment (xezon) [inline src/game/common/rts/specialpower.cpp:133]: `power->Set_Is_Allocated();` missing?

File: src/game/common/rts/specialpower.cpp [hunk 3]
```diff
@@ -62,6 +62,98 @@ bool SpecialPowerStore::Can_Use_Special_Power(Object *obj, const SpecialPowerTem
...
+            power->Set_Name_ID(str, ++g_theSpecialPowerStore->m_nextSpecialPowerID);
+            power->Set_Is_Allocated();
+            g_theSpecialPowerStore->m_specialPowerTemplates.push_back(power);
+        }
+    } else {
+        if (power != nullptr) {
+            throw CODE_06;
+        }
+
+        power = new SpecialPowerTemplate();
```
Comment (xezon) [inline src/game/common/rts/specialpower.cpp:125]: Is allocating here and above its checking that `power` is not null. Error?

File: src/game/common/rts/specialpower.cpp [hunk 4]
```diff
@@ -62,6 +62,98 @@ bool SpecialPowerStore::Can_Use_Special_Power(Object *obj, const SpecialPowerTem
...
+        ini->Init_From_INI(power, SpecialPowerTemplate::Get_Field_Parse());
+    }
+}
+
+SpecialPowerTemplate::SpecialPowerTemplate() :
+    m_id(0),
+    m_type(SPECIAL_INVALID),
+    m_reloadTime(0),
+    m_requiredScience(SCIENCE_INVALID),
+    m_detectionTime(300),
```
Comment (xezon) [inline src/game/common/rts/specialpower.cpp:148]: `m_academyClassify` is uninitialized.

File: src/game/common/rts/specialpower.cpp [hunk 5]
```diff
@@ -77,3 +169,25 @@ const SpecialPowerTemplate *SpecialPowerStore::Find_Special_Power_Template_By_ID
 
     return nullptr;
 }
+
+static const char *s_academyClassificationTypeNames[] = { "ACT_NONE", "ACT_UPGRADE_RADAR", "ACT_SUPERPOWER", nullptr };
+
+// clang-format off
+const FieldParse SpecialPowerTemplate::s_specialPowerFieldParse[] = {
+
```
Comment (xezon) [inline src/game/common/rts/specialpower.cpp]: Remove blank line.

File: src/game/common/rts/specialpower.h [hunk 6]
```diff
@@ -65,8 +77,15 @@ class SpecialPowerStore : public SubsystemInterface
     virtual void Update() override {}
 
     bool Can_Use_Special_Power(Object *obj, const SpecialPowerTemplate *special_power_template);
+    SpecialPowerTemplate *Find_Special_Power_Template_Private(Utf8String name);
+    const SpecialPowerTemplate *Find_Special_Power_Template(Utf8String name);
```
Comment (xezon) [inline src/game/common/rts/specialpower.h:81]: This function and 2 below can be const.

File: src/game/common/rts/specialpower.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/common/rts/specialpower.cpp:133]: No, it specifically doesn't exist (I had it there previously and it caused things to break until I realized my mistake).

File: src/game/common/rts/specialpower.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/common/rts/specialpower.cpp:125]: That's checking the value of power obtained from the earlier call to Find_Special_Power_Template_Private.

File: src/game/common/rts/specialpower.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/common/rts/specialpower.cpp:177]: fixed.

File: src/game/common/rts/specialpower.h (hunk 6 above)
Comment (jonwil) [reply src/game/common/rts/specialpower.h:81]: Original is not const. Want to keep things matching original at this point (obviously as the project progresses, we can deviate from original in these areas but I see no benefit to doing it now.)

File: src/game/common/rts/specialpower.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/common/rts/specialpower.cpp:90]: Original returns int.

File: src/game/common/rts/specialpower.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/common/rts/specialpower.cpp:148]: fixed.

File: src/game/common/rts/specialpower.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/common/rts/specialpower.cpp]: fixed.

**Review by xezon — APPROVED** [review]
Special Powers work as usual.

**Pull request merged**